source/s2n/s2n_tls_channel_handler.c

```diff
@@ -573,6 +588,11 @@ static int s_s2n_handler_process_read_message(
         } else {
             aws_mem_release(outgoing_read_message->allocator, outgoing_read_message);
         }
+
+        if (blocked == S2N_NOT_BLOCKED) {
+            /* read is complete */
+            break;
+        }
```

Comment:
why is this here? you don't need to read the blocked status at all

Reply:
Thanks for reviewing. We allocate a buffer before every call to recv. This is just an optimization that allows us to break out of the loop early if the read is complete. This way, we avoid an unnecessary recv that would return a read less than 0 with blocked status, allowing us to break out of the loop.

Reply:
If you want to make sure the s2n-tls buffer has been drained, it would be much less fragile to use `s2n_peek`

```suggestion
        if (s2n_peek(conn) == 0) {
            /* the s2n-tls buffer has been completely processed */
            break;
        }
```

That being said... do you know what will register epoll interest for future reads? I'm not familiar with CRT but you usually drain the socket until you get blocked and then register epoll at that point.

Reply:
But in general, if you're receiving a lot of data from the socket it's better to loop and eat the extra allocation over waiting for epoll to wake you up for every single record.